### PR TITLE
[CARBONDATA-691] After Compaction records count are mismatched.

### PR DESCRIPTION
**Pull request by sounakr**
**Problem** : After Compaction record count mismatches with actual count. 

**Analysis** :The Partitioning method of compaction was wrong. In getPartition method of CarbonScanRDD.scala supposed to make a list all the blocks of all the segments that needs to be merged and then make the partition based on taskNo.  Then each partitioned list is given to each executor. But currently after partitioning the complete list of blocks are being send to each executor for merging. As each executors merging all the blocks of all the segments, multiple executors doubles the data. 

**Fix** : Fix the getPartition method logic to process proper list of blocks to executors.
             Fix Horizontal Partitioning which merged with IUD.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/917/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala [hunk 1]
```diff
@@ -255,77 +256,75 @@ class CarbonMergerRDD[K, V](
...
     for (eachSeg <- carbonMergerMapping.validSegments) {
-      // map for keeping the relation of a task and its blocks.
-      val taskIdMapping: java.util.Map[String, java.util.List[TableBlockInfo]] = new
-            java.util.HashMap[String, java.util.List[TableBlockInfo]]
 
       // map for keeping the relation of a task and its blocks.
       job.getConfiguration.set(CarbonInputFormat.INPUT_SEGMENT_NUMBERS, eachSeg)
 
+      val updateDetails: UpdateVO = updateStatusManager.getInvalidTimestampRange(eachSeg)
+      // val updateStatusManager = carbonLoadModel.getSegmentUpdateStatusManager
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: remove commented code

**Comment by jackylk** [comment]
Is there a testcase to reproduce this bug and verify this PR?

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
+/*
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: It seems some problem with commit, this class is shown as new class

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    this.carbonBlockInfoList = carbonSplitListInfo;
+  }
+
+  @Override public String[] getLocations() {
+    Set<String> locations = new HashSet<String>();
+    for (CarbonInputSplit splitInfo : carbonBlockInfoList) {
+      try {
+        locations.addAll(Arrays.asList(splitInfo.getLocations()));
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: Add logger or throw the exception

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        for (String node : split.getLocations()) {
+          Integer nodeOccurence = nodeAndOccurenceMapping.get(node);
+          if (null == nodeOccurence) {
+            nodeAndOccurenceMapping.put(node, 1);
+          } else {
+            nodeOccurence++;
+          }
+        }
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: Add logger or throw the exception

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java (hunk 2 above)
Comment (sounakr) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: This is a new class. Could you please tell me what is the problem with commit?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1001/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala (hunk 1 above)
Comment (sounakr) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: Done. Removed the commented code.

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala [hunk 5]
```diff
@@ -235,121 +238,141 @@ class CarbonMergerRDD[K, V](
...
+        dataFileFooter = CarbonUtil.readMetadatFile(
+          CarbonInputSplit.getTableBlockInfo(splits))
+      } catch {
+        case e: IOException =>
+          logError("Exception in preparing the data file footer for compaction " + e.getMessage)
+          throw e
+      }
+
+      // Calculate the Cardinality of the new segment
+      var targetCardinality = dataFileFooter.getSegmentInfo.getColumnCardinality
```
Comment (ravikiran23) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: move the target cardinality calculation to a new method

**Comment by sounakr** [comment]
@jackylk : This problem will get reproduce in a multi node system where a single Load will create multiple carbon data file is by multiple tasks. I have tested the code changes in a multi node cluster with large data and output ic coming correctly. 
Another problem observed while testing is choosing the correct cardinality of the new segment formed due to compaction. Previously we used to copy the last segment cardinality to the new compacted segment. But with IUD features updates can run on any segments and in case the updates happens on any intermediate segments then there is high probability that cardinality of that segment will be more than the last segment. So now instead of copying the cardinality from the last segment we calculate the highest cardinility values from all the segments.

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java (hunk 3 above)
Comment (sounakr) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: Done. Throwing exception now.

Comment (sounakr) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: Done. Throwing Exception.

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java (hunk 4 above)
Comment (sounakr) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonInputSplitTaskInfo.java]: Done. Throwing Exception.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1006/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1012/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala (hunk 5 above)
Comment (sounakr) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: Done. Moved the logic of cardinality calculation to a function.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1013/

**Comment by ravipesala** [comment]
LGTM

**Pull request merged**